File: Binary_Trees/2677-cousins-in-binary-tree-ii/cousins-in-binary-tree-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque

class Solution:
    def replaceValueInTree(self, root: TreeNode) -> TreeNode:
        if root is None:
            return root

        temp = deque([root])
        root.val = 0

        while temp:
            lsize = len(temp)
            total_sum = 0

            cousins = {}
            for _ in range(lsize):
                node = temp.popleft()
                child_sum = 0

                if node.left is not None:
                    temp.append(node.left)
                    child_sum += node.left.val

                if node.right is not None:
                    temp.append(node.right)
                    child_sum += node.right.val

                cousins[node] = child_sum
                total_sum += child_sum

            for node, child_sum in cousins.items():
                if node.left is not None:
                    node.left.val = total_sum - child_sum

                if node.right is not None:
                    node.right.val = total_sum - child_sum

        return root

File: Binary_Trees/2677-cousins-in-binary-tree-ii/cousins-in-binary-tree-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* replaceValueInTree(TreeNode* root) {
        if (root == NULL)
            return root;

        queue<TreeNode*> temp;
        temp.push(root);

        root->val = 0;

        while (!temp.empty()) {
            int lsize = temp.size();
            int totalSum = 0;

            unordered_map<TreeNode*, int> cousins;
            for (int i = 0; i < lsize; i++) {
                TreeNode* node = temp.front();
                temp.pop();
                int childsum = 0;

                if (node->left != NULL) {
                    temp.push(node->left);
                    childsum += node->left->val;
                }

                if (node->right != NULL) {
                    temp.push(node->right);
                    childsum += node->right->val;
                }

                cousins[node] = childsum;
                totalSum += childsum;
            }

            for (auto& x : cousins) {
                TreeNode* node = x.first;
                int childsum = x.second;

                if (node->left != NULL)
                    node->left->val = totalSum - childsum;

                if (node->right != NULL)
                    node->right->val = totalSum - childsum;
            }
        }

        return root;
    }
};